src/background_workers.c

```diff
@@ -245,130 +308,126 @@ void *RedisAI_Run_ThreadMain(void *arg) {
-          //               available yet
-          // dag_device_complete: all DAG operations for the current device have completed,
-          //                      there's nothing more this worker can do
-          // dag_all_devices_complete: all DAG operations for any device have completed,
-          //                           which means that the computation of the DAG is complete
-          RedisAI_DagRunSessionStep(batch_rinfo[0], run_queue_info->devicestr,
-                                    &dag_progress, &dag_device_complete, &dag_all_devices_complete);
-        } else {
-          // Run the (potentially batched) model or the (non-batched) script run
-          RAI_ModelRunScriptRunSession(batch_rinfo);
+      // For better readability, we create a variable that says if the current
+      // run will actually be batched or not
+      int batched_run = array_len(batch_rinfo) > 1;
+
+      // Variable holding the fact that running triggered an error.
+      int run_error = 0;
+      // Run the computation step (batched or not)
+      if (do_run == 1) {
+        // We're done with the queue here, items have been evicted so we can
+        // safely unlock the queue mutex, to allow other threads to operate
+        // on the same queue. The evicted items at this point are only visible
+        // to this worker.
+        pthread_mutex_unlock(&run_queue_info->run_queue_mutex);
+
+        // For simplicity, we call into different functions whether the run
+        // is batched or not
+        if (batched_run == 1) {
+          RedisAI_BatchedDagRunSessionStep(batch_rinfo, run_queue_info->devicestr);
+        }
+        else {
+          RedisAI_DagRunSessionStep(batch_rinfo[0], run_queue_info->devicestr);
+        }
+
+        // Lock the queue again: we're done operating on evicted items only, we need
+        // to update the queue with the new information after run
+        pthread_mutex_lock(&run_queue_info->run_queue_mutex);
+
+        // Run is over, now iterate over the run info structs in the batch
+        // and see if any error was generated
+        int dagError = 0;
+        int dagRefCount = -1;
+        for (long long i=0; i<array_len(batch_rinfo); i++) {
+          RedisAI_RunInfo *rinfo = batch_rinfo[i];
+          // We lock on the DAG because error and refcount could be set from
+          // other threads operating on the same DAG
+          pthread_mutex_lock(rinfo->dagMutex);
+          dagError = *rinfo->dagError;
+          dagRefCount = *rinfo->dagRefCount;
+          pthread_mutex_unlock(rinfo->dagMutex);
+
+          // We record that there was an error for later on
+          run_error = dagError;
+
+          // If there was an error and the reference count for the dag
+          // has gone to zero and the client is still around, we unblock
+          if (dagError && dagRefCount == 0 && rinfo->client) {
+            RedisModule_UnblockClient(rinfo->client, rinfo);
```

Comment:
here is a suggestion:
Have a function - `Dag_DecreseRefCount` or `Dag_Free`, or `Dag_CleanUp` or similar
This will atomically reduce the ref count, and call the `RedisModule_UnblockClient`

In further step, I think this function will call a callback, to enable async running. By default (meaning, by sending Redis command) this callback will call `RedisModule_UnblockClient`